services/reminders/reminders.py

```diff
@@ -0,0 +1,128 @@
+                + (
+                    session.query(ReminderOnSpectrum)
+                    .where(ReminderOnSpectrum.number_of_reminders != 0)
+                    .all()
+                )
+                + (
+                    session.query(ReminderOnGCN)
+                    .where(ReminderOnGCN.number_of_reminders != 0)
+                    .all()
+                    + (
+                        session.query(ReminderOnShift)
+                        .where(ReminderOnShift.number_of_reminders != 0)
+                        .all()
+                    )
+                )
+            )
+
+            for reminder in reminders:
+                await self.put([reminder.id, reminder.__class__])
+
+    async def service(self):
+        while True:
+
+            reminder_id, reminder_type = await queue.get()
+
+            with DBSession() as session:
+                reminder = session.query(reminder_type).get(reminder_id)
+                dt = Time(datetime.utcnow(), format='datetime') - Time(
+                    reminder.next_reminder, format='datetime'
+                )
+                if dt.jd > 0:
+                    await self.put([reminder_id, reminder_type])
+                else:
+                    print(f"Sending reminder {reminder.id}")
+
+                    if reminder_type == Reminder:
+                        text_to_send = (
+                            f"Reminder of source *{reminder.obj_id}*: {reminder.text}"
+                        )
+                        url_endpoint = f"/source/{reminder.obj_id}"
+                    elif reminder_type == ReminderOnSpectrum:
+                        text_to_send = f"Reminder of spectrum *{reminder.spectrum_id}*: {reminder.text}"
+                        url_endpoint = f"/source/{reminder.spectrum_id}"
+                    elif reminder_type == ReminderOnGCN:
+                        text_to_send = f"Reminder of GCN event *{reminder.gcnevent_id}*: {reminder.text}"
+                        url_endpoint = f"/gcn_events/{reminder.gcnevent_id}"
+                    elif reminder_type == ReminderOnShift:
+                        text_to_send = (
+                            f"Reminder of *shift {reminder.shift_id}*: {reminder.text}"
+                        )
+                        url_endpoint = "/shifts"
+                    else:
+                        return self.error(f'Unknown reminder type "{reminder_type}".')
+
+                    session.add(
+                        UserNotification(
+                            user=reminder.user,
+                            text=text_to_send,
+                            notification_type="mention",
+                            url=url_endpoint,
+                        )
+                    )
+
+                    flow = Flow()
+                    flow.push('*', "skyportal/FETCH_NOTIFICATIONS", {})
+
+                    reminder.number_of_reminders = reminder.number_of_reminders - 1
```

Comment:
Couldn't we simply delete the reminder from the DB when it reaches 0 ?

Reply:
I think we usually keep DB entries like for completed requests.